Premium

YusukeShimizu opened the pull request:
This PR adds premium to peerswap.
[design doc](https://docs.google.com/document/d/1enPKthscJCRX_bG_12fwcUjfm9-FdTBoRvPIL1HLR1k/edit#heading=h.uoxht6u5sf15)

## Usage
### doing swap
A premium limit will be added to the existing swap command.
If not set, default is 0.

#### cln
```bash
lightning-cli peerswap-swap-out [short channel id] [amount in sats] [asset: btc or lbtc] [premium limit in sats]
```

#### lnd
```bash
pscli swapout --sat_amt=[sat amount] --channel_id=[chan_id from above] --asset=btc --premium_limit=[premium limit]
```

### configure premium rate
The premium rate is pert per million and can be set for swapin and swapout respectively.
`swap_amount * premium_rate_ppm / 1000000` is the premium.
Negative values can also be set to give a discount

For usage instructions, please refer to [YusukeShimizu/peerswap@premium-feature/docs/usage.md#premium](https://github.com/YusukeShimizu/peerswap/blob/premium-feature/docs/usage.md#premium).

The gRPC interface can be found here: [YusukeShimizu/peerswap@premium-feature/peerswaprpc/peerswaprpc.proto#L35-L42](https://github.com/YusukeShimizu/peerswap/blob/premium-feature/peerswaprpc/peerswaprpc.proto#L35-L42).

### examples
If the max_acceptable_premium is lower than responder.
```sh
./bin/pscli peerswap1 swapout --channel_id $CHANID --sat_amt 1000000 --asset btc --premium_limit 1
2023/12/23 02:02:36 rpc error: code = Unknown desc = from the LND peer: unacceptable premium: 100, limit: 1
```

success
```sh
./bin/pscli peerswap1 swapout --channel_id $CHANID --sat_amt 1000000 --asset btc --premium_limit 100
{
  "swap": {
    "id": "eb5f41a515f737956300a38ab1f5286a67d580c4d3f838afdf48537c1133ed69",
    "created_at": "1702532654",
    "asset": "btc",
    "type": "swap-out",
    "role": "sender",
    "state": "State_SwapOutSender_AwaitTxConfirmation",
    "initiator_node_id": "02ea1fd791bb1b639e7119ac54505e3f553a7bb942877ccbb359affa8547046988",
    "peer_node_id": "034066fecfcdad7cb2f002c8c4234dbab26688db9617509ede4c7372c79ab94d0d",
    "amount": "1000000",
    "channel_id": "110:1:1",
    "opening_tx_id": "83f0b1174a3c55934ca9fa60398dbdfa76394db1450cbe2c1d2369ef80095b15",
    "claim_tx_id": "",
    "cancel_message": "",
    "lnd_chan_id": "120946279120897",
    "premium": "100"
  }
}
```

wtogami commented:
Asking for opinions: Should we have fine grained configuration to allow for setting certain peers a different premium fee rate?
I suppose that would complicate things a lot and make the configuration look ugly. But is this a highly desirable feature?

Impa10r commented:
> Asking for opinions: Should we have fine grained configuration to allow for setting certain peers a different premium fee rate? I suppose that would complicate things a lot and make the configuration look ugly. But is this a highly desirable feature?

I think no. But allow separate premiums for received swap ins and swap outs, including negative values. This will allow managing liquidity  among BTC, LBTC and Lightning.

YusukeShimizu commented:
> Asking for opinions: Should we have fine grained configuration to allow for setting certain peers a different premium fee rate?
I suppose that would complicate things a lot and make the configuration look ugly. But is this a highly desirable feature?

In the future, it may be better to allow premiums to be set according to the degree of use, such as channels that flow easily in both directions or channels that flow only in one direction.
However, since this has nothing to do with protocol changes, why not implement this in this PR and check and consider it later?

YusukeShimizu commented:
>  But allow separate premiums for received swap ins and swap outs, including negative values. This will allow managing liquidity among BTC, LBTC and Lightning.

It is possible to configure swap in and swap out in `policy.conf`, including negative values.

grubles commented:
> Asking for opinions: Should we have fine grained configuration to allow for setting certain peers a different premium fee rate? I suppose that would complicate things a lot and make the configuration look ugly. But is this a highly desirable feature?

Yes I think so. Channels may have different characteristics that would benefit from having individual premium/discount rates to incentivize swapping in one direction or the other, similar to how routing fee rates can be used to influence how a channel is used?

grubles commented:
Would it be useful to have the premium rate limit be a config option in `policy.conf` too?

grubles commented:
Hitting this with CLN as sender:

```
$ lightning-cli --signet peerswap-swap-out 140490x1x0 100000000 lbtc 1000000
{
   "code": -32603,
   "message": "-32603: Incompatible types: SwapOut.bool (bool) != float64"
}
```

grubles commented:
CLN plugin crashed when trying a `listswaps`:

```
$ lightning-cli --signet peerswap-listswaps
{
   "code": -4,
   "message": "Plugin terminated before replying to RPC call."
}
```

Looks like `peerswapd` segfaulted when trying `listswaps` too:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xdb1460]

goroutine 142825 [running]:
github.com/elementsproject/peerswap/swap.(*SwapData).GetPremium(...)
        /home/user/builds/master_peerswap/swap/swap.go:239
github.com/elementsproject/peerswap/peerswaprpc.PrettyprintFromServiceSwap(0xc0002c5600)
        /home/user/builds/master_peerswap/peerswaprpc/server.go:576 +0x300
github.com/elementsproject/peerswap/peerswaprpc.(*PeerswapServer).ListSwaps(0xf6cf00?, {0xc0006f4120?, 0x498fe6?}, 0x0?)
        /home/user/builds/master_peerswap/peerswaprpc/server.go:321 +0xfd
github.com/elementsproject/peerswap/peerswaprpc._PeerSwap_ListSwaps_Handler({0x10505a0?, 0xc000400840}, {0x12b8dd8, 0xc0006f40f0}, 0xc00043e620, 0x0)
        /home/user/builds/master_peerswap/peerswaprpc/peerswaprpc_grpc.pb.go:357 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000320b40, {0x12c0880, 0xc00040e820}, 0xc0006e6360, 0xc00001f8f0, 0x199ffc8, 0x0)
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:1318 +0xb18
google.golang.org/grpc.(*Server).handleStream(0xc000320b40, {0x12c0880, 0xc00040e820}, 0xc0006e6360, 0x0)
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:1659 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:955 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 142824
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:953 +0x246
```

The logs appear to show that the swap I attempted succeeded however. 


```
2024/01/26 19:03:15 [INFO] [Swap:cdd8bff7205897ad5ed687e830c434a31abf80fdca1f011d32e256bbab836057] Swap claimed with preimage 6fb9a8eb04df189ffe3f78772ff74b29c36ab55ef3923edf5a6b56999cd8cc09
```

YusukeShimizu commented:
Thanks for the test.
Perhaps some of the behavior is a bug.
I'll investigate and fix it.

YusukeShimizu commented:
FIXED.

> [premium info is missing from the CLN listpeers](https://github.com/ElementsProject/peerswap/pull/264#issuecomment-1912361102)

SwapInPremiumRatePPM and SwapInPremiumRatePPM are added to the ListPeers.  
We are currently managing the peerswapd and the cln-plugin in duplicate, but this will be fixed by https://github.com/ElementsProject/peerswap/pull/231.

> [Hitting this with CLN as sender](https://github.com/ElementsProject/peerswap/pull/264#issuecomment-1912381165)

This was handled by rearranging the order of commands.
I think it is natural that the force flag should be at the end.

> [CLN plugin crashed when trying a listswaps](https://github.com/ElementsProject/peerswap/pull/264#issuecomment-1912619329).

The source of the information was an error, which has been corrected.

grubles commented:
At 9c34c47 my `peerswapd` segfaults when attempting a swap-out. 

```
2024/01/31 15:31:22 [INFO] [Swap:69d53116566e72ac131f343a7dfc492879042c02752c0557369cbd8674703eee] Start new swap-out: peer: 036e680823a129ac9a2eeb8c43514a5b6cb09ae28e3e698a0e34aaa8865d6cc344 chanId: 140490:1:0 initiator: 035ca2fe4793a5e789ce846062eb4834f573c060d9200ce77544a29b48a0aa5923 amount 100000000
2024/01/31 15:31:22 [DEBUG] [FSM] event:id: 69d53116566e72ac131f343a7dfc492879042c02752c0557369cbd8674703eee, Event_ActionSucceeded on State_SwapOutSender_CreateSwap
2024/01/31 15:31:22 [DEBUG] [FSM] event:id: 69d53116566e72ac131f343a7dfc492879042c02752c0557369cbd8674703eee, Event_ActionSucceeded on State_SwapOutSender_SendRequest
2024/01/31 15:31:22 [DEBUG] [Messenger] From: 036e680823a129ac9a2eeb8c43514a5b6cb09ae28e3e698a0e34aaa8865d6cc344 got msgtype: a45b for swap: 69d53116566e72ac131f343a7dfc492879042c02752c0557369cbd8674703eee
2024/01/31 15:31:22 [DEBUG] [FSM] event:id: 69d53116566e72ac131f343a7dfc492879042c02752c0557369cbd8674703eee, Event_OnFeeInvoiceReceived on State_SwapOutSender_AwaitAgreement
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xcb6c9f]

goroutine 69 [running]:
github.com/elementsproject/peerswap/swap.(*CheckPremiumAmount).Execute(0xc0006bd600?, 0x10ad8bb?, 0xc000318780)
        /home/user/builds/master_peerswap/swap/actions.go:215 +0x11f
github.com/elementsproject/peerswap/swap.(*SwapStateMachine).SendEvent(0xc0006bd600, {0x10a2a12, 0x1a}, {0x12b42d0?, 0xc0001b8e00})
        /home/user/builds/master_peerswap/swap/fsm.go:233 +0x7ae
github.com/elementsproject/peerswap/swap.(*SwapService).OnSwapOutAgreementReceived(0xc00055c2c0?, 0xc0001b8e00)
        /home/user/builds/master_peerswap/swap/service.go:695 +0x92
github.com/elementsproject/peerswap/swap.(*SwapService).OnMessageReceived(0xc0001cfee0?, {0xc000841c20, 0x42}, {0xc0005f5028, 0x4}, {0xc00055c2c0, 0x2b9, 0x2c0})
        /home/user/builds/master_peerswap/swap/service.go:218 +0x906
github.com/elementsproject/peerswap/lnd.(*MessageListener).Start.func1()
        /home/user/builds/master_peerswap/lnd/messagelistener.go:84 +0x473
created by github.com/elementsproject/peerswap/lnd.(*MessageListener).Start in goroutine 1
        /home/user/builds/master_peerswap/lnd/messagelistener.go:60 +0x10b
```

YusukeShimizu commented:
I fixed it on https://github.com/ElementsProject/peerswap/pull/264/commits/d8b683d478332794947be52c42e84fccd470ee89

YusukeShimizu commented:
I fixed it on https://github.com/ElementsProject/peerswap/pull/264/commits/2b9dfa303d09fba18da9b510e318d1813f45b408 same as swapout

grubles commented:
After editing `policy.conf` to change the swap premium rates while CLN / peerswap are running, CLN peerswap does not output the correct values for the premium rates when running `peerswap-reloadpolicy`. In this example the correct values would be `swap_in_premium_rate_ppm` = 40000 and `swap_out_premium_rate_ppm` rate = 60000: 

```
$ lightning-cli --signet peerswap-reloadpolicy                                                                                  
{
   "reserve_onchain_msat": 0,
   "allowlisted_peers": [
      "035ca2fe4793a5e789ce846062eb4834f573c060d9200ce77544a29b48a0aa5923"
   ],
   "suspicious_peers": [],
   "accept_all_peers": false,
   "min_swap_amount_msat": 100000000,
   "allow_new_swaps": true,
   "swap_in_premium_rate_ppm": 0,     <------
   "swap_out_premium_rate_ppm": 0    <------
}
```

Though the remote peer receives a poll from the CLN peerswap with the correct new rates:

```
2024/02/21 00:09:22 [DEBUG] Received poll from peer 036e680823a129ac9a2eeb8c43514a5b6cb09ae28e3e698a0e34aaa8865d6cc344: {"version":5,"assets":["lbtc","btc"],"peer_allowed":true,"swap_in_premium_rate_ppm":40000,"swap_out_premium_rate_ppm":60000}
```



For LND peerswap, when doing the same thing as above, the premium rates aren't included in the json output at all:

```
$ ./builds/peerswap/out/pscli reloadpolicy
{
  "reserve_onchain_msat":  "0",
  "min_swap_amount_msat":  "100000000",
  "accept_all_peers":  false,
  "allow_new_swaps":  true,
  "allowlisted_peers":  [
    "036e680823a129ac9a2eeb8c43514a5b6cb09ae28e3e698a0e34aaa8865d6cc344"
  ],
  "suspicious_peer_list":  []
}
```

YusukeShimizu commented:
https://github.com/ElementsProject/peerswap/pull/264#issuecomment-1955572416

When the reload policy was executed, the configuration values were updated, but they were not being displayed, which was causing this behavior.

Definitions are being added so that they are displayed.
https://github.com/ElementsProject/peerswap/pull/264/commits/82d3121a03aa94cc37c6517267c06f823466f348

tsjk commented:
It would make sense to have different premiums for different assets too. And, for different peers as well. With defaults to fall back.

Impa10r commented:
The required premium rates for in and out swaps should be advertised by the receiver in its response to REQUEST_POLL message. So that the initiator can see these rates in `listpeers` output. 

Attempting swaps blindly with 'max acceptable premium rate' is a no go. Peerswap is already suffering from uncertainty about available balances, additional uncertainty about fees will kill it. People hate attempting and seeing constant probing in swaps history.

tsjk commented:
> The required premium rates for in and out swaps should be advertised by the receiver in its response to REQUEST_POLL message. So that the initiator can see these rates in `listpeers` output.
> 
> Attempting swaps blindly with 'max acceptable premium rate' is a no go. Peerswap is already suffering from uncertainty about available balances, additional uncertainty about fees will kill it. People hate attempting and seeing constant probing in swaps history.

I totally agree here!

SwapMarket commented:
Hello. Is this PR abandoned, or still in the works? No one keeps any Liquid balance because it can be snatched cheaply. There is a need for a swap out premium.

YusukeShimizu commented:
Thank you for the feedback.
This PR is in progress.

YusukeShimizu commented:
I have made it possible to set premiums individually for BTC and LBTC.  
Additionally, it is now possible to configure settings on a per-peer basis.  
You can set it up as shown below.

```
reserve_onchain_msat=0
allowlisted_peers=peer1
allowlisted_peers=peer2
allowlisted_peers=peer3
accept_all_peers=0
# swap_amount * premium_rate_ppm / 1000000 is the premium
# Negative values can also be set to give a discount
# * The value specified by `btc_swap_in_premium_rate_ppm` is the premium rate for BTC swap in
# * The value specified by `btc_swap_out_premium_rate_ppm` is the premium rate for BTC swap out
# * The value specified by `lbtc_swap_in_premium_rate_ppm` is the premium rate for LBTC swap in
# * The value specified by `lbtc_swap_out_premium_rate_ppm` is the premium rate for LBTC swap out
[base_premium_rate]
btc_swap_in_premium_rate_ppm=100
# Set the premium rate for each peer
# * The value specified by `peer1.btc_swap_in_premium_rate_ppm` is the premium rate for BTC swap in for peer1
# * The value specified by `peer1.btc_swap_out_premium_rate_ppm` is the premium rate for BTC swap out for peer1
# * The value specified by `peer1.lbtc_swap_in_premium_rate_ppm` is the premium rate for LBTC swap in for peer1
# * The value specified by `peer1.lbtc_swap_out_premium_rate_ppm` is the premium rate for LBTC swap out for peer1
[peers_premium_rate]
peer1.btc_swap_in_premium_rate_ppm=1000
peer2.btc_swap_out_premium_rate_ppm=2000
```

Impa10r commented:
@YusukeShimizu 

1. Will we set these params via API/cli calls or must edit the policy.conf directly (poor UX)?
2. What is `reserve_onchain_msat`? Keep min balance for onchain BTC? Any chance to have one for LBTC as well?

YusukeShimizu commented:
The current code only supports direct editing of `policy.conf`.  
However, it is also possible to make edits, such as modifying the allow list, through API or CLI calls.  
I plan to add this functionality.

The `reserve_onchain_msat` setting was my mistake and has been removed as it was an unnecessary and meaningless configuration.  
Although it was defined as a parameter in peerswap, it was not used in the logic.

Impa10r commented:
I did not specify anything in policy yet, but listpeers shows 10000 ppm for my other node. I think 0 is a better default.

```
 "premium": {
        "btc_swap_in_premium_rate_ppm": "10000",
        "btc_swap_out_premium_rate_ppm": "10000",
        "lbtc_swap_in_premium_rate_ppm": "10000",
        "lbtc_swap_out_premium_rate_ppm": "10000"
      }
```

Impa10r commented:
after updating the policy file, it is necessary to restart `peerswapd` for the changes to take effect. `pscli reloadpolicy` is not enough despite it shows the updated values?

YusukeShimizu commented:
It seems that setting the base premium to 0 would indeed be better in terms of compatibility issues. I will make the correction.

Although the internal values are being updated due to the reload policy, it seems that it is taking some time for these changes to be reflected. I believe there is an advantage in addressing this because, with immediate polling when the premium is updated, the changes will be reflected quickly for list peers and the latest premium will also be sent to peers. Therefore, I will take action on this.

YusukeShimizu commented:
In summary, I am planning the following updates:
1. Change the default value of premium settings to 0ppm
2. Enable updating premium via API or CLI calls
3. Implement immediate polling when updating premium
4. Allow displaying both your own and the counterparty's premium in listpeer

YusukeShimizu commented:
I have implemented https://github.com/ElementsProject/peerswap/pull/264#issuecomment-2415541187.

To make it easier to update the premium, I have updated the architecture from using policy.conf to using a key-value store for configuration. For usage instructions, please refer to https://github.com/YusukeShimizu/peerswap/blob/premium-feature/docs/usage.md#premium.

The gRPC interface can be found here: https://github.com/YusukeShimizu/peerswap/blob/premium-feature/peerswaprpc/peerswaprpc.proto#L35-L42.

tsjk commented:
> I have implemented [#264 (comment)](https://github.com/ElementsProject/peerswap/pull/264#issuecomment-2415541187).
> 
> To make it easier to update the premium, I have updated the architecture from using policy.conf to using a key-value store for configuration. For usage instructions, please refer to https://github.com/YusukeShimizu/peerswap/blob/premium-feature/docs/usage.md#premium.
> 
> The gRPC interface can be found here: https://github.com/YusukeShimizu/peerswap/blob/premium-feature/peerswaprpc/peerswaprpc.proto#L35-L42.

Swap-Outs also support a premium limit parameter, I suppose.

grubles commented:
The premium docs for CLN seem to use the command arguments:

```
$ ./builds/lightning/cli/lightning-cli --signet peerswap-getdefaultpremiumrate --asset lbtc --operation swap_in
./builds/lightning/cli/lightning-cli: --asset: unrecognized option
```

Omitting the `--asset --operation` works though.

```
$ ./builds/lightning/cli/lightning-cli --signet peerswap-getdefaultpremiumrate lbtc swap_out
0
```

YusukeShimizu commented:
The documentation contains an error.
For the CLN (Core Lightning Network) case, it’s correct to omit the parameter.
https://github.com/ElementsProject/peerswap/pull/264/commits/4795258d538473edb86a3764b19fde9328aadadc

Could you try setting the premium limit and testing it, as shown in the example command for CLN’s ⁠swapin?
https://github.com/YusukeShimizu/peerswap/blob/premium-feature/docs/usage.md#swap-in

Swap-in Command Example.
```sh
$ ./builds/peerswap/out/pscli swapin --channel_id 248103699313721344 --sat_amt 10000001 --asset lbtc 10000
```

YusukeShimizu commented:
I've rebased onto the latest master and removed the draft status.

grubles commented:
Can we make the `getdefaultpremiumrate`  `--asset` and `--operation` flags not require all caps? Is there a reason to have them be that way?

```
$ ./builds/peerswap/out/pscli getdefaultpremiumrate --asset lbtc --operation swap_in  
2025/02/14 13:55:40 rpc error: code = Unknown desc = invalid asset type: ASSET_UNSPECIFIED
```

YusukeShimizu commented:
I've updated it so that you can specify the flags regardless of uppercase or lowercase.
https://github.com/ElementsProject/peerswap/pull/264/commits/178522abe397fe8b48acc01b7b3035413930fc36

grubles commented:
It seems setting the premium rates in `policy.conf` is ignored:

```
$ cat ~/.lightning/signet/peerswap/policy.conf 
swap_in_premium_rate_ppm=40000
swap_out_premium_rate_ppm=60000
```

```
$ ./builds/lightning/cli/lightning-cli --signet peerswap-getdefaultpremiumrate lbtc swap_out
1000
```

grubles commented:
When updating the `swap_out` premium for LND peerswap, this occurs:

```
$ ./builds/peerswap/out/pscli updatedefaultpremiumrate --asset lbtc --operation swap_out --rate 2000
2025/02/18 14:46:25 rpc error: code = Unavailable desc = error reading from server: EOF

LND logs:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xcdb360]

goroutine 1104 [running]:
github.com/elementsproject/peerswap/premium.(*PremiumRate).Asset(0xa79edc?)
        /home/user/builds/peerswap/premium/premium.go:69
github.com/elementsproject/peerswap/premium.(*BBoltPremiumStore).SetDefaultRate.(*BBoltPremiumStore).SetRate.func1(0x0?)
        /home/user/builds/peerswap/premium/store.go:46 +0x7b
go.etcd.io/bbolt.(*DB).Update(0xc0015732c0?, 0xc0000c79b0)
        /home/user/go/pkg/mod/go.etcd.io/bbolt@v1.3.6/db.go:741 +0x75
github.com/elementsproject/peerswap/premium.(*BBoltPremiumStore).SetRate(...)
        /home/user/builds/peerswap/premium/store.go:41
github.com/elementsproject/peerswap/premium.(*BBoltPremiumStore).SetDefaultRate(0xcdb32b?, 0x1115668?)
        /home/user/builds/peerswap/premium/store.go:77 +0x4d
github.com/elementsproject/peerswap/premium.(*Setting).SetDefaultRate(0xc0004a6100, 0xc0000c7a68?)
        /home/user/builds/peerswap/premium/premium.go:181 +0x1e
github.com/elementsproject/peerswap/peerswaprpc.(*PeerswapServer).UpdateDefaultPremiumRate(0xc000bd6780, {0x10d3940?, 0xc00184fab8?}, 0xc00186ede0)
        /home/user/builds/peerswap/peerswaprpc/server.go:624 +0xfd
github.com/elementsproject/peerswap/peerswaprpc._PeerSwap_UpdateDefaultPremiumRate_Handler({0x10d3940, 0xc000bd6780}, {0x1344ad0, 0xc00186edb0}, 0xc0002de310, 0x0)
        /home/user/builds/peerswap/peerswaprpc/peerswaprpc_grpc.pb.go:673 +0x1a9
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000000f00, {0x134cfc0, 0xc0019181a0}, 0xc001876360, 0xc000c7ebd0, 0x1a8d298, 0x0)
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:1318 +0xaea
google.golang.org/grpc.(*Server).handleStream(0xc000000f00, {0x134cfc0, 0xc0019181a0}, 0xc001876360, 0x0)
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:1659 +0x9da
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:955 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 1139
        /home/user/go/pkg/mod/google.golang.org/grpc@v1.50.0/server.go:953 +0x226
```

YusukeShimizu commented:
Thank you for feedback https://github.com/ElementsProject/peerswap/pull/264#issuecomment-2666789895.
Fixes on https://github.com/ElementsProject/peerswap/pull/264/commits/455eeb4e59ee2d94dfb7eb1ebf2d2712c6cbc59a

grubles commented:
Looks like the unacceptable premium rate check for LND peerswap is calculating the PPM incorrectly? 
```
[user@peerswap ~]$ ./builds/lightning/cli/lightning-cli --signet peerswap-setdefaultpremiumrate lbtc swap_out 1001
{}
[user@peerswap ~]$ ./builds/peerswap/out/pscli swapout --channel_id 248103699313721344 --sat_amt 10000001 --asset lbtc --premium_limit 1000
2025/02/21 11:06:59 rpc error: code = Unknown desc = from the CLN peer: unacceptable premium: 10010, limit: 1000
```

YusukeShimizu commented:
Because premium_limit=1001 is interpreted as the upper limit in sat, the actually calculated premium (approximately 10,010 sat) exceeds that limit, resulting in the error “unacceptable premium: 10010, limit: 1001.”  
This behavior is functioning as intended. If you prefer to specify the limit in ppm, you may need to adjust the configuration or use different parameters.

Would specifying it by rate be more convenient?

grubles commented:
Ok I was confused thinking the `premium_limit` was the rate but it's really the total premium amount. 

> Would specifying it by rate be more convenient?

That's a good question. What if it was the premium rate, but peerswap asked you to confirm Y/N to the total premium amount for the swap?

Impa10r commented:
> Actually on second thought would a confirmation request break UIs that have implemented support for peerswap, like RTL or peerswap-web? cc @Impa10r

Don't think Y/N interaction is a good idea. It will probably require support by pscli as well. If the rate limit is set by the user as ppm, that would be enough.

grubles commented:
Ok @YusukeShimizu let's go with a PPM limit instead of total premium amount limit. That way PPM is the only unit the node operator needs to worry about?

YusukeShimizu commented:
I changed the Premium limit setting from sat to a ppm-based rate.  
https://github.com/ElementsProject/peerswap/pull/264/commits/c84e53af90fa6a330f749c7aa6b1b79f526de611

Impa10r commented:
> I changed the Premium limit setting from sat to a ppm-based rate. [c84e53a](https://github.com/ElementsProject/peerswap/commit/c84e53af90fa6a330f749c7aa6b1b79f526de611)

Can rates go negative?

YusukeShimizu commented:
Yes, they can go negative.

grubles commented:
Tested ACK [c84e53a](https://github.com/ElementsProject/peerswap/pull/264/commits/c84e53af90fa6a330f749c7aa6b1b79f526de611)

Impa10r commented:
Is it intended that if a peer's rate was not set, but the default one was, `getpeerpremiumrate` returns that default rate. but after the first time `updatepeerpremiumrate` was initialized in the db, `getpeerpremiumrate` starts returning that specific rate? This is a bit confusing.

I think if peer's specific rate was untouched it should return 0.

Impa10r commented:
I observe default rate not propagating to peers:

```
~ $ pscli listpeers
{
  "peers": [
    {
      "node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "swaps_allowed": true,
      "supported_assets": [
        "btc",
        "lbtc"
      ],
      "channels": [
        {
          "channel_id": "2796274673227464705",
          "local_balance": "788955",
          "remote_balance": "210100",
          "active": true
        }
      ],
      "as_sender": {
        "swaps_out": "6",
        "swaps_in": "12",
        "sats_out": "1500838",
        "sats_in": "3412992"
      },
      "as_receiver": {
        "swaps_out": "6",
        "swaps_in": "11",
        "sats_out": "2584121",
        "sats_in": "4266216"
      },
      "paid_fee": "19847",
      "peer_premium": {
        "node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
        "rates": [
          {
            "asset": "BTC",
            "operation": "SWAP_IN",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "BTC",
            "operation": "SWAP_OUT",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "LBTC",
            "operation": "SWAP_IN",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "LBTC",
            "operation": "SWAP_OUT",
            "premium_rate_ppm": "0"
          }
        ]
      }
    }
  ]
}
~ $ pscli2 getdefaultpremiumrate --asset BTC --operation SWAP_IN
{
  "asset": "BTC",
  "operation": "SWAP_IN",
  "premium_rate_ppm": "4"
}
```

After I have set a specific value, the peer saw that immediately.

Impa10r commented:
```
I tried a swap with a large negative premium rate, I don't see it being accounted for:

~ $ pscli listactiveswaps
{
  "swaps": [
    {
      "id": "d54f62e3381eb57f574b8c6b4f6ba7dff3fc2e2ba5642ae68ba5a16587e9ef13",
      "created_at": "1741890698",
      "asset": "lbtc",
      "type": "swap-in",
      "role": "receiver",
      "state": "State_SwapInReceiver_AwaitTxConfirmation",
      "initiator_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "peer_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "amount": "100000",
      "channel_id": "2543197:37:1",
      "opening_tx_id": "410b1eec087bea5d97e1308e3b5331edd4a50a21a037ef9456f019142d9ce09c",
      "claim_tx_id": "",
      "cancel_message": "",
      "lnd_chan_id": "2796274673227464705",
      "premium_amount": "0"
    }
  ]
}

~ $ pscli2 listactiveswaps
{
  "swaps": [
    {
      "id": "d54f62e3381eb57f574b8c6b4f6ba7dff3fc2e2ba5642ae68ba5a16587e9ef13",
      "created_at": "1741890698",
      "asset": "lbtc",
      "type": "swap-in",
      "role": "sender",
      "state": "State_SwapInSender_AwaitClaimPayment",
      "initiator_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "peer_node_id": "020ed4ea09c56c3e398190ae5ac60504fe4e56d6989d49e2f7845f14608c546854",
      "amount": "100000",
      "channel_id": "2543197:37:1",
      "opening_tx_id": "410b1eec087bea5d97e1308e3b5331edd4a50a21a037ef9456f019142d9ce09c",
      "claim_tx_id": "",
      "cancel_message": "",
      "lnd_chan_id": "2796274673227464705",
      "premium_amount": "0"
    }
  ]
}

~ $ pscli2 listpeers
{
  "peers": [
    {
      "node_id": "020ed4ea09c56c3e398190ae5ac60504fe4e56d6989d49e2f7845f14608c546854",
      "swaps_allowed": true,
      "supported_assets": [
        "btc",
        "lbtc"
      ],
      "channels": [
        {
          "channel_id": "2796274673227464705",
          "local_balance": "410100",
          "remote_balance": "588955",
          "active": true
        }
      ],
      "as_sender": {
        "swaps_out": "7",
        "swaps_in": "18",
        "sats_out": "2684121",
        "sats_in": "5437662"
      },
      "as_receiver": {
        "swaps_out": "19",
        "swaps_in": "26",
        "sats_out": "3572151",
        "sats_in": "5971563"
      },
      "paid_fee": "975",
      "peer_premium": {
        "node_id": "020ed4ea09c56c3e398190ae5ac60504fe4e56d6989d49e2f7845f14608c546854",
        "rates": [
          {
            "asset": "BTC",
            "operation": "SWAP_IN",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "BTC",
            "operation": "SWAP_OUT",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "LBTC",
            "operation": "SWAP_IN",
            "premium_rate_ppm": "0"
          },
          {
            "asset": "LBTC",
            "operation": "SWAP_OUT",
            "premium_rate_ppm": "-10000"
          }
        ]
      }
    }
  ]
}

```

Impa10r commented:
for a positive preimum rate, both counterparties see the same report, so not possible to understand who paid the premium to whom: 

```
~ $ pscli listactiveswaps
{
  "swaps": [
    {
      "id": "027e62d4cc1bd94a7fa69e47e76fec6cf708cd9c89a504ba8084c94024c9b6c0",
      "created_at": "1741891106",
      "asset": "lbtc",
      "type": "swap-out",
      "role": "receiver",
      "state": "State_SwapOutReceiver_AwaitClaimInvoicePayment",
      "initiator_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "peer_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "amount": "100000",
      "channel_id": "2543197:37:1",
      "opening_tx_id": "1659a308e9f0ee524172052d6790bdbec3d271ed8860728202e5bd9d5676677d",
      "claim_tx_id": "",
      "cancel_message": "",
      "lnd_chan_id": "2796274673227464705",
      "premium_amount": "-1000"
    }
  ]
}
~ $ pscli2 listactiveswaps
{
  "swaps": [
    {
      "id": "027e62d4cc1bd94a7fa69e47e76fec6cf708cd9c89a504ba8084c94024c9b6c0",
      "created_at": "1741891106",
      "asset": "lbtc",
      "type": "swap-out",
      "role": "sender",
      "state": "State_SwapOutSender_AwaitTxConfirmation",
      "initiator_node_id": "02e74acffd6375fa71a038e780064ae1765ebaf520bf7b9a5681e8ff9bfe36ce29",
      "peer_node_id": "020ed4ea09c56c3e398190ae5ac60504fe4e56d6989d49e2f7845f14608c546854",
      "amount": "100000",
      "channel_id": "2543197:37:1",
      "opening_tx_id": "1659a308e9f0ee524172052d6790bdbec3d271ed8860728202e5bd9d5676677d",
      "claim_tx_id": "",
      "cancel_message": "",
      "lnd_chan_id": "2796274673227464705",
      "premium_amount": "-1000"
    }
  ]
}

```

grubles commented:
```
   "premium_amount": "-1000"
```

For this we could maybe change it to something like `premium_rate_ppm` and then have a separate `total_premium_amount` to show the actual premium paid? ?

Impa10r commented:
> ```
>    "premium_amount": "-1000"
> ```
> 
> For this we could maybe change it to something like `premium_rate_ppm` and then have a separate `total_premium_amount` to show the actual premium paid? ?

ppm is not needed, it can be computed from premium_amount and swap amount. I think the premium amount is already correct, except the sign. it should be positive for the peer that received premium, as negative for the one who paid it.

YusukeShimizu commented:
I just added a new RPC and CLI command to remove a peer-specific premium rate, so it falls back to the global default once the per-peer rate is deleted. 

Also, I adjusted the sign in PrettyprintFromServiceSwap for the sender role to ensure it’s displayed correctly.

https://github.com/ElementsProject/peerswap/pull/264/commits/b26c621553e1d2fd0a751cdb94ddfce8db8460dd

### delete premium rate
#### lnd
```sh
./bin/pscli peerswap1 updatepremiumrate --asset LBTC --operation swap_out --node_id 02149a96f829dcfbea2089fd5e3fe3fd9d2d08157694102d3ddf6db346d5dd161d --rate 100
{
  "asset": "LBTC",
  "operation": "SWAP_OUT",
  "premium_rate_ppm": "100"
}
```
```sh
./bin/pscli peerswap1 deletepeerpremiumrate --asset LBTC --operation SWAP_OUT --node_id 02149a96f829dcfbea2089fd5e3fe3fd9d2d08157694102d3ddf6db346d5dd161d
{
  "asset": "LBTC",
  "operation": "SWAP_OUT",
}
```
fall back to default
```sh
./bin/pscli peerswap1 getpeerpremiumrate --asset LBTC --operation SWAP_OUT --node_id 02149a96f829dcfbea2089fd5e3fe3fd9d2d08157694102d3ddf6db346d5dd161d
{
  "asset": "LBTC",
  "operation": "SWAP_OUT",
  "premium_rate_ppm": "10000"
}
```

#### cln
```sh
./bin/clncli peerswap-setpremiumrate 02bc29431b3a13172dd1c469aaa718990c37b3a3a4b8098523ad4c3c61f7fb282c lbtc SWAP_OUT 10000
{
   "asset": 2,
   "operation": 2,
   "premium_rate_ppm": 100
}
```
```sh
./bin/clncli peerswap-deletepremiumrate 02bc29431b3a13172dd1c469aaa718990c37b3a3a4b8098523ad4c3c61f7fb282c lbtc SWAP_OUT
{
   "asset": 2,
   "operation": 2
}
```
fall back to default
```sh
./bin/clncli peerswap-getpremiumrate 02bc29431b3a13172dd1c469aaa718990c37b3a3a4b8098523ad4c3c61f7fb282c lbtc SWAP_OUT
{
   "asset": 2,
   "operation": 2,
   "premium_rate_ppm": 10000
}
```

### PrettyprintFromServiceSwap
```sh
./bin/pscli peerswap1 swapout --channel_id $CHANID --sat_amt 1000000 --asset lbtc --premium_limit_rate_ppm 100000
{
  "swap": {
    "id": "c7a54d60bdf3deb0b0e77bc074cc625d84d5e62dda7aa97df40edf5c9e1a0f31",
    "created_at": "1741931535",
    "asset": "lbtc",
    "type": "swap-out",
    "role": "sender",
    "state": "State_SwapOutSender_AwaitTxConfirmation",
    "initiator_node_id": "03854c727bacb9a39da5cb5bd50fdf250747d32b11c4ff56389adfecabf5c91445",
    "peer_node_id": "0276b3477147a0374e5d372c53c394bd1e286cbba613a7ff0fc8afce2d6def07e3",
    "amount": "1000000",
    "channel_id": "116:1:1",
    "opening_tx_id": "05f3d7a7929758ab4c0def7919da41f9b4c184cbd47dd2f85b3694f25a561529",
    "claim_tx_id": "",
    "cancel_message": "",
    "lnd_chan_id": "127543348887553",
    "premium_amount": "-10000"
  }
}
```
```sh
./bin/pscli peerswap2 listswaps
{
  "swaps": [
    {
      "id": "c7a54d60bdf3deb0b0e77bc074cc625d84d5e62dda7aa97df40edf5c9e1a0f31",
      "created_at": "1741931535",
      "asset": "lbtc",
      "type": "swap-out",
      "role": "receiver",
      "state": "State_SwapOutReceiver_AwaitClaimInvoicePayment",
      "initiator_node_id": "03854c727bacb9a39da5cb5bd50fdf250747d32b11c4ff56389adfecabf5c91445",
      "peer_node_id": "03854c727bacb9a39da5cb5bd50fdf250747d32b11c4ff56389adfecabf5c91445",
      "amount": "1000000",
      "channel_id": "116:1:1",
      "opening_tx_id": "05f3d7a7929758ab4c0def7919da41f9b4c184cbd47dd2f85b3694f25a561529",
      "claim_tx_id": "",
      "cancel_message": "",
      "lnd_chan_id": "127543348887553",
      "premium_amount": "10000"
    }
  ]
}
```

Impa10r commented:
Tested ACK https://github.com/ElementsProject/peerswap/pull/264/commits/b26c621553e1d2fd0a751cdb94ddfce8db8460dd

YusukeShimizu commented:
We have renamed “Premium” to “Global.”   https://github.com/ElementsProject/peerswap/pull/264/commits/4d37f4da5dec102ac37a8fabe07f5c381aa4cb2c
Below are the default values for the Global Premium:

| Asset | Swap Type | Default PPM |
| ----- | --------- | ----------- |
| BTC   | swap in   | 0           |
| LBTC  | swap in   | 0           |
| BTC   | swap out  | 2000        |
| LBTC  | swap out  | 1000        |

https://github.com/ElementsProject/peerswap/pull/264/commits/6369d6a2545ef66d84157a366fbe911dc9da1400

Impa10r commented:
ACK https://github.com/ElementsProject/peerswap/commit/6369d6a2545ef66d84157a366fbe911dc9da1400

grubles commented:
```
$ lightning-cli --signet peerswap-getglobalpremiumrate btc swap_in
{
   "asset": 1,
   "operation": 1
}
```

When the global rate is zero, the `premium_rate_ppm` field is missing?

YusukeShimizu commented:
It was caused by how `proto.Message` was marshaled in a non-human-readable manner.  
I addressed it by specifying the marshal option to ensure the field is included.
[26268a0 – Use human-readable marshal option](https://github.com/ElementsProject/peerswap/pull/264/commits/26268a07e923019f9ef46fa5932d183db41de1dc).

I’ll keep this in mind for the changes introduced in [PR #231](https://github.com/ElementsProject/peerswap/pull/231) as well.

grubles commented:
Tested ACK [26268a0](https://github.com/ElementsProject/peerswap/commit/26268a07e923019f9ef46fa5932d183db41de1dc)
